File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-doc-5-4/VersionedContractId.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class VersionedContractId : ISerialized
    {
        public VersionedContractId(XmlNode xmlNode)
        {
            XmlNode contractIdNode = xmlNode.SelectSingleNode("contractId");
            
            if (contractIdNode != null)
            {
                if (contractIdNode.Attributes["href"] != null || contractIdNode.Attributes["id"] != null) 
                {
                    if (contractIdNode.Attributes["id"] != null) 
                    {
                        contractIdIDRef_ = contractIdNode.Attributes["id"].Value;
                        ContractId ob = new ContractId(contractIdNode);
                        IDManager.SetID(contractIdIDRef_, ob);
                    }
                    else if (contractIdNode.Attributes["href"] != null)
                    {
                        contractIdIDRef_ = contractIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        contractId_ = new ContractId(contractIdNode);
                    }
                }
                else
                {
                    contractId_ = new ContractId(contractIdNode);
                }
            }
            
        
            XmlNode versionNode = xmlNode.SelectSingleNode("version");
            
            if (versionNode != null)
            {
                if (versionNode.Attributes["href"] != null || versionNode.Attributes["id"] != null) 
                {
                    if (versionNode.Attributes["id"] != null) 
                    {
                        versionIDRef_ = versionNode.Attributes["id"].Value;
                        XsdTypeNonNegativeInteger ob = new XsdTypeNonNegativeInteger(versionNode);
                        IDManager.SetID(versionIDRef_, ob);
                    }
                    else if (versionNode.Attributes["href"] != null)
                    {
                        versionIDRef_ = versionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        version_ = new XsdTypeNonNegativeInteger(versionNode);
                    }
                }
                else
                {
                    version_ = new XsdTypeNonNegativeInteger(versionNode);
                }
            }
            
        
            XmlNode effectiveDateNode = xmlNode.SelectSingleNode("effectiveDate");
            
            if (effectiveDateNode != null)
            {
                if (effectiveDateNode.Attributes["href"] != null || effectiveDateNode.Attributes["id"] != null) 
                {
                    if (effectiveDateNode.Attributes["id"] != null) 
                    {
                        effectiveDateIDRef_ = effectiveDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(effectiveDateNode);
                        IDManager.SetID(effectiveDateIDRef_, ob);
                    }
                    else if (effectiveDateNode.Attributes["href"] != null)
                    {
                        effectiveDateIDRef_ = effectiveDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        effectiveDate_ = new IdentifiedDate(effectiveDateNode);
                    }
                }
                else
                {
                    effectiveDate_ = new IdentifiedDate(effectiveDateNode);
                }
            }
            
        
        }
        
    
        #region ContractId_
        private ContractId contractId_;
        public ContractId ContractId_
        {
            get
            {
                if (this.contractId_ != null)
                {
                    return this.contractId_; 
                }
                else if (this.contractIdIDRef_ != null)
                {
                    contractId_ = IDManager.getID(contractIdIDRef_) as ContractId;
                    return this.contractId_; 
                }
                else
                {
                      return this.contractId_; 
                }
            }
            set
            {
                if (this.contractId_ != value)
                {
                    this.contractId_ = value;
                }
            }
        }
        #endregion
        
        public string contractIdIDRef_ { get; set; }
        #region Version_
        private XsdTypeNonNegativeInteger version_;
        public XsdTypeNonNegativeInteger Version_
        {
            get
            {
                if (this.version_ != null)
                {
                    return this.version_; 
                }
                else if (this.versionIDRef_ != null)
                {
                    version_ = IDManager.getID(versionIDRef_) as XsdTypeNonNegativeInteger;
                    return this.version_; 
                }
                else
                {
                      return this.version_; 
                }
            }
            set
            {
                if (this.version_ != value)
                {
                    this.version_ = value;
                }
            }
        }
        #endregion
        
        public string versionIDRef_ { get; set; }
        #region EffectiveDate_
        private IdentifiedDate effectiveDate_;
        public IdentifiedDate EffectiveDate_
        {
            get
            {
                if (this.effectiveDate_ != null)
                {
                    return this.effectiveDate_; 
                }
                else if (this.effectiveDateIDRef_ != null)
                {
                    effectiveDate_ = IDManager.getID(effectiveDateIDRef_) as IdentifiedDate;
                    return this.effectiveDate_; 
                }
                else
                {
                      return this.effectiveDate_; 
                }
            }
            set
            {
                if (this.effectiveDate_ != value)
                {
                    this.effectiveDate_ = value;
                }
            }
        }
        #endregion
        
        public string effectiveDateIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-doc-5-4/VersionedContractId.cpp
// VersionedContractId.cpp 
#include "VersionedContractId.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

VersionedContractId::VersionedContractId(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //contractIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* contractIdNode = xmlNode->FirstChildElement("contractId");

   if(contractIdNode){contractIdIsNull_ = false;}
   else{contractIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- contractIdNode , address : " << contractIdNode << std::endl;
   #endif
   if(contractIdNode)
   {
      if(contractIdNode->Attribute("href") || contractIdNode->Attribute("id"))
      {
          if(contractIdNode->Attribute("id"))
          {
             contractIdIDRef_ = contractIdNode->Attribute("id");
             contractId_ = boost::shared_ptr<ContractId>(new ContractId(contractIdNode));
             contractId_->setID(contractIdIDRef_);
             IDManager::instance().setID(contractIdIDRef_,contractId_);
          }
          else if(contractIdNode->Attribute("href")) { contractIdIDRef_ = contractIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { contractId_ = boost::shared_ptr<ContractId>(new ContractId(contractIdNode));}
   }

   //versionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* versionNode = xmlNode->FirstChildElement("version");

   if(versionNode){versionIsNull_ = false;}
   else{versionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- versionNode , address : " << versionNode << std::endl;
   #endif
   if(versionNode)
   {
      if(versionNode->Attribute("href") || versionNode->Attribute("id"))
      {
          if(versionNode->Attribute("id"))
          {
             versionIDRef_ = versionNode->Attribute("id");
             version_ = boost::shared_ptr<XsdTypeNonNegativeInteger>(new XsdTypeNonNegativeInteger(versionNode));
             version_->setID(versionIDRef_);
             IDManager::instance().setID(versionIDRef_,version_);
          }
          else if(versionNode->Attribute("href")) { versionIDRef_ = versionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { version_ = boost::shared_ptr<XsdTypeNonNegativeInteger>(new XsdTypeNonNegativeInteger(versionNode));}
   }

   //effectiveDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* effectiveDateNode = xmlNode->FirstChildElement("effectiveDate");

   if(effectiveDateNode){effectiveDateIsNull_ = false;}
   else{effectiveDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- effectiveDateNode , address : " << effectiveDateNode << std::endl;
   #endif
   if(effectiveDateNode)
   {
      if(effectiveDateNode->Attribute("href") || effectiveDateNode->Attribute("id"))
      {
          if(effectiveDateNode->Attribute("id"))
          {
             effectiveDateIDRef_ = effectiveDateNode->Attribute("id");
             effectiveDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(effectiveDateNode));
             effectiveDate_->setID(effectiveDateIDRef_);
             IDManager::instance().setID(effectiveDateIDRef_,effectiveDate_);
          }
          else if(effectiveDateNode->Attribute("href")) { effectiveDateIDRef_ = effectiveDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { effectiveDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(effectiveDateNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<ContractId> VersionedContractId::getContractId()
{
   if(!this->contractIdIsNull_){
        if(contractIdIDRef_ != ""){
             return boost::shared_static_cast<ContractId>(IDManager::instance().getID(contractIdIDRef_));
        }else{
             return this->contractId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ContractId>();
   }
}
boost::shared_ptr<XsdTypeNonNegativeInteger> VersionedContractId::getVersion()
{
   if(!this->versionIsNull_){
        if(versionIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeNonNegativeInteger>(IDManager::instance().getID(versionIDRef_));
        }else{
             return this->version_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeNonNegativeInteger>();
   }
}
boost::shared_ptr<IdentifiedDate> VersionedContractId::getEffectiveDate()
{
   if(!this->effectiveDateIsNull_){
        if(effectiveDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(effectiveDateIDRef_));
        }else{
             return this->effectiveDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
}

